### PR TITLE
feat: Add defensive replay protection check into WriteAcknowledgement

### PR DESCRIPTION
**Pull request by hoangdv2429**
## Description

closes: #5740

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by hoangdv2429** [comment]
I think the e2e test is flaking again, can someone rerun it

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @hoangdv2429. I think the original intention of the issue is to add the check in [04-channel's `WriteAcknowledgement`](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/packet.go#L284), so that all middleware/applications can benefit from it. If we add the check only in ics29 middleware, then only the fee middleware will be protected against the replay attack.

**Comment by hoangdv2429** [comment]
> Thank you, @hoangdv2429. I think the original intention of the issue is to add the check in [04-channel's `WriteAcknowledgement`](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/packet.go#L284), so that all middleware/applications can benefit from it. If we add the check only in ics29 middleware, then only the fee middleware will be protected against the replay attack.

Got it sir. I'll update it

**Comment by crodriguezvega** [comment]
Thanks for changing the code, @hoangdv2429. It would fantastic if you could also a test. You can take as example the test that was added for the same check in `RecvPacket` [here](https://github.com/cosmos/ibc-go/pull/5651/files#diff-ce312850968149b915fa31bf5158acb380a7efe3931048052422b0c6bd92ab80R607-R621). Thank you!

File: modules/core/04-channel/keeper/packet_test.go [hunk 1]
```diff
@@ -777,6 +777,22 @@ func (suite *KeeperTestSuite) TestWriteAcknowledgement() {
...
 			false,
 		},
+		{
+			"packet already received",
+			func() {
+				path.Setup()
+
+				sequence, err := path.EndpointA.SendPacket(defaultTimeoutHeight, disabledTimeoutTimestamp, ibctesting.MockPacketData)
+				suite.Require().NoError(err)
+				packet = types.NewPacket(ibctesting.MockPacketData, 1, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID, defaultTimeoutHeight, disabledTimeoutTimestamp)
```
Comment (damiannolan) [inline modules/core/04-channel/keeper/packet_test.go:787]: should we use the returned `sequence` from the initial packet send? 

```suggestion
				packet = types.NewPacket(ibctesting.MockPacketData, sequence, path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID, defaultTimeoutHeight, disabledTimeoutTimestamp)
```

**Review by damiannolan — APPROVED** [review]
Looks good! LGTM! Thank you @hoangdv2429 🙏🏻

File: modules/core/04-channel/keeper/packet.go [hunk 2]
```diff
@@ -305,6 +305,14 @@ func (k Keeper) WriteAcknowledgement(
 		)
 	}
 
+	// REPLAY PROTECTION: Even after the packet has been processed and acknowledgement is returned,
+	// it won't be possible to replay the acknowledgement processing on the sender side as the packet
+	// commitment won't exist, thus this check is simply to protect unnecessary state from being written into storage.
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/packet.go:310]:
```suggestion
	// REPLAY PROTECTION: The recvStartSequence will prevent historical proofs from allowing replay
	// attacks on packets processed in previous lifecycles of a channel. After a successful channel
	// upgrade all packets under the recvStartSequence will have been processed and thus should be
	// rejected. Any asynchronous acknowledgement writes from packets processed in a previous lifecycle of a channel
	// will also be rejected.
```

**Review by colin-axner — APPROVED** [review]
Thanks @hoangdv2429! LGTM, I left a suggested improvement for the in code documentation

**Pull request merged**